File: c++/355. Design Twitter.cpp
struct Tweet{
    int id;
    int order;  
    Tweet(int id_, int order_):id(id_), order(order_) {
        
    }
};

class Twitter {
public:
    /** Initialize your data structure here. */
    Twitter() {
        order = 0;
    }
    
    /** Compose a new tweet. */
    void postTweet(int userId, int tweetId) {
        user_tweets[userId].push_back(Tweet(tweetId, order++));
    }
    
    /** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
    vector<int> getNewsFeed(int userId) {
        vector<int> res;
        vector<Tweet> user_t;
        if (user_tweets.count(userId)) {
            vector<Tweet> v = user_tweets[userId];
            user_t = getLastTen(v);
        }
        
        if (friends.count(userId)) {
            for (int fid : friends[userId]) {
                if (user_tweets.count(fid)) {
                    vector<Tweet> v = user_tweets[fid];
                    user_t.insert(user_t.end(), v.begin(), v.end());
                }
            }
        }
        
        sort(user_t.begin(), user_t.end(), [](const Tweet &t1, const Tweet &t2){
            return t1.order > t2.order;    
        });
        
        vector<Tweet> v = getFirstTen(user_t);
        for (int i = 0; i < v.size(); i++)
            res.push_back(v[i].id);
        return res;
    }
    
    /** Follower follows a followee. If the operation is invalid, it should be a no-op. */
    void follow(int followerId, int followeeId) {
        if (followerId == followeeId) return;
        friends[followerId].insert(followeeId);
    }
    
    /** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
    void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId) return;
        friends[followerId].erase(followeeId);
    }
    
private:
    vector<Tweet> getFirstTen(vector<Tweet> &tweets) {
        int last = 10;
        last = min((int)tweets.size(), last);
        return vector<Tweet>(tweets.begin(), tweets.begin() + last);
    }
    
    vector<Tweet> getLastTen(vector<Tweet> &tweets) {
        int last = 10;
        last = min((int)tweets.size(), last);
        return vector<Tweet>(tweets.end() - last, tweets.end());
    }

    unordered_map<int, vector<Tweet>> user_tweets;
    unordered_map<int, set<int>> friends;
    int order;
};

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * vector<int> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */

File: java/src/main/java/code/355. Design Twitter.java
package code;

import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

class Twitter {

    private static int timeStamp = 0;

    private static class User {
        private int id;
        public Set<Integer> followed;
        public Tweet head;

        public User(int id) {
            this.id = id;
            this.head = null;
            this.followed = new HashSet<>();
            this.followed.add(this.id);
        }

        public void follow(int userId) {
            this.followed.add(userId);
        }

        public void unfollow(int userId) {
            if (this.id!=userId)
                this.followed.remove(userId);
        }

        public void post(int tweetId) {
            Tweet twt = new Tweet(tweetId, timeStamp);
            timeStamp++;
            twt.next = head;
            head = twt;
        }

    }

    private static class Tweet {
        private int id;
        private int time;
        private Tweet next;

        public Tweet(int id, int time) {
            this.id = id;
            this.time = time;
        }
    }

    private Map<Integer, User> userMap = new HashMap<>();

    /** Initialize your data structure here. */
    public Twitter() {

    }

    /** Compose a new tweet. */
    public void postTweet(int userId, int tweetId) {
        if (!userMap.containsKey(userId)) {
            userMap.put(userId, new User(userId));
        }
        User user = userMap.get(userId);
        user.post(tweetId);
    }

    /** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
    public List<Integer> getNewsFeed(int userId) {
        if (!userMap.containsKey(userId)) {
            userMap.put(userId, new User(userId));
        }
        User user = userMap.get(userId);
        Set<Integer> users = user.followed;

        PriorityQueue<Tweet> pq = new PriorityQueue<>(users.size(), (a, b)->(b.time - a.time));
        for (int followeeId: user.followed) {
            Tweet twt = userMap.get(followeeId).head;
            if (twt==null) {
                continue;
            }
            pq.add(twt);
        }

        List<Integer> res = new ArrayList<>();
        while (!pq.isEmpty()) {
            if (res.size()==10) {
                break;
            }
            Tweet twt = pq.poll();
            if (twt.next!=null) {
                pq.add(twt.next);
            }
            res.add(twt.id);
        }
        return res;

    }

    /** Follower follows a followee. If the operation is invalid, it should be a no-op. */
    public void follow(int followerId, int followeeId) {
        if (!userMap.containsKey(followerId)) {
            userMap.put(followerId, new User(followerId));
        }
        if (!userMap.containsKey(followeeId)) {
            userMap.put(followeeId, new User(followeeId));
        }
        User user = userMap.get(followerId);
        user.follow(followeeId);
    }

    /** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
    public void unfollow(int followerId, int followeeId) {
        if (userMap.containsKey(followerId)) {
            User user = userMap.get(followerId);
            user.unfollow(followeeId);
        }

    }
}
/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */